[fix] huge parser performance boost for reading big attribute values

MathiasWP opened the pull request:
## The bug
I worked on a svelte-kit application with some very large SVG's, and they made it almost impossible to work on the application (because of slow development time).

I have created a repo with one of the SVG's if you want to see it yourself: https://github.com/MathiasWP/svelte-attribute-bottleneck

Note that i created the repo with the old svelte-template because i wanted to ensure this wasn't a Vite bug.

## The solution
I found out that the method `read_attribute_value`  + `read_sequence` in `src/compiler/parse/state/tag.ts` was the big bottleneck. It has a while loop that loops `n = template.length` amount of times (in the test cases i have added this was 1.7 million times). 

On every loop a RegEx check was ran with the parser. This RegEx call was suboptimal, so i changed the code from using `parser.match_regex` to using `parser.match` with characters that mirror the RegEx call. 

Changing from using RegEx to a strict string equality check massively improved the performance:

With regex: _~30s_
With new setup:  _~260ms_

This is a 99% shave-off in execution time 🎉 

### Before submitting the PR, please make sure you do the following
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [X] Prefix your PR title with `[feat]`, `[fix]`, `[chore]`, or `[docs]`.
- [X] This message body should clearly illustrate what problems it solves.
- [X] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [X] Run the tests with `npm test` and lint the project with `npm run lint`

benmccann commented on src/compiler/parse/state/tag.ts:
```diff
@@ -160,8 +165,9 @@ export default function tag(parser: Parser) {
+	
 	const unique_names: Set<string> = new Set();
-
+	
```
no need to add whitespace on this line. I also probably remove lots of the new lines being added in this PR
```suggestion

```

benmccann commented on test/runtime/samples/input-crazy-value-attribute/_config.js:
```diff
@@ -0,0 +1,10 @@
+
+	test({ assert, target }) {
+		const input = target.querySelector('input');
+		assert.equal(input.value.length > 100_000, true);
```
numeric separators were added in Node 12.5 and we support older versions, which is why the CI is failing
```suggestion
		assert.equal(input.value.length > 100000, true);
```

MathiasWP replied:
Ahh, makes sense! I'll resolve this, and also increase the value because the value is much longer than 100 000 characters

benmccann commented:
wow, that's a pretty nice speedup!

I didn't look super closely at the test, but I wasn't sure exactly what it was testing or how it worked at first glance. Is it a test that would fail without this change? Performance improvements are hard to add tests for...

MathiasWP commented:
> wow, that's a pretty nice speedup!
> 
> 
> 
> I didn't look super closely at the test, but I wasn't sure exactly what it was testing or how it worked at first glance. Is it a test that would fail without this change? Performance improvements are hard to add tests for...

@benmccann i added three tests, one with an svg, one with an input and one with a textarea. In all of the tests they have an attribute that is crazy long (it's originally a base64 encoded string of one of the svg's from the initial problem). 

Tbh i wasn't sure how to test for performance either, so i initially only created the tests and was going to file an issue pointing to the tests. But i found out that if i added logs in the src code, then they would be logged realtime in the tests. So i divide and conquered my way through the src code with logs to find out where the big bottleneck was.

The tests will fail without the proposed changes because they will trigger a timeout error. They took around 30s to run. You can see this if you replace my optimization with the current solution 👍🏼

Conduitry commented:
I'm not wild about adding huge amounts of text to the repo to test this. Can the tests be written so that the large amounts of text that are parsed are generated at runtime?

In either case, whatever tests we keep, the `.solo` on them needs to be removed.

MathiasWP commented:
@Conduitry i changed the test so it generates the value on runtime, removed `.solo` and removed 2/3 tests.

tanhauhau commented on src/compiler/parse/state/tag.ts:
```diff
@@ -465,15 +471,10 @@ function read_attribute_value(parser: Parser) {
-
 	let value;
 	try {
-		value = read_sequence(parser, () => !!parser.match_regex(regex));
```
just curious, why is this written as a spread, instead of defining another constant list at the top of the file?

MathiasWP replied:
No special reason to be honest. Do you want med to define everything as a constant?

tanhauhau replied:
yea that would be better

tanhauhau commented on test/runtime/samples/input-crazy-value-attribute/main.svelte:
```diff
@@ -0,0 +1,5 @@
+	export let value;
+</script>
+
+<input type="text" {value} />
```
i dont think this will test whatever you are trying to test, since, there's the props are passed in during runtime, and the changes you made is on the parser which runs in the compile time.

i dont think we can have a test case over here to test how fast the parsing is, but we can test the correctness of the changes.

maybe can have a test case that has all the possible whitespace characters in between attribute values and make sure the parser can parse it correctly.

anyway since this change is more of a refactoring / behavior change, if existing test case does not break then the change should be safe to make.

MathiasWP replied:
Thank you for clearing this up, @tanhauhau! I can create a test with all possible white space characters in between if you want. Should i do it?

tanhauhau replied:
yup that would be great. I'll probably run the test before and after the change, and see if the change introduce any  regression

MathiasWP replied:
I am wondering if this test will be a little weird, because it seems like the whitespace characters that were in the original RegEx are invalid whitespace characters for attribute values. So I'm not sure if this test will make sense. Please correct me if I'm wrong @tanhauhau

tanhauhau replied:
its okay to be without adding new tests. since this PR is not fixing any bugs or changing any behavior

ota-meshi commented on src/compiler/parse/state/tag.ts:
```diff
@@ -9,6 +9,9 @@ import read_script from '../read/script';
 import { closing_tag_omitted, decode_character_references } from '../utils/html';
 
+// characters equivalent to using \s in regex
+const whitespace_characters = ['\r', '\n', '\t', '\f', '\v', ' ', '\u00a0', '\u1680', '\u2000-', '\u200a', '\u2028', '\u2029', '\u202f', '\u205f', '\u3000', '\ufeff'];
```
Probably a mistake.

```suggestion
const whitespace_characters = ['\r', '\n', '\t', '\f', '\v', ' ', '\u00a0', '\u1680', '\u2000', '\u2001', '\u2002', '\u2003', '\u2004', '\u2005', '\u2006', '\u2007', '\u2008', '\u2009', '\u200a', '\u2028', '\u2029', '\u202f', '\u205f', '\u3000', '\ufeff'];
```

MathiasWP replied:
I got this list of characters from regex101:

<img width="416" alt="Screenshot 2022-07-10 at 09 56 03" src="https://user-images.githubusercontent.com/48158184/178136305-9a97c7b7-8470-43fd-9738-da64e5171b77.png">

I was also unsure about the "-" being correct, so if it is a mistake I'll remove it!

MathiasWP replied:
You can see the "-" on the end at the \u2000 part, which is why i included it

ota-meshi replied:
You can learn the meaning of hyphens.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes

MathiasWP replied:
Oh, i did not see all the extra characters you added. Thank you for clearing this up!

MathiasWP commented:
Question about the array new array that i have created: Is its usage similar to the `invalid_attribute_name_character` in the `src/runtime/internal/ssr.ts` file?
<img width="1265" alt="Screenshot 2022-07-10 at 12 48 53" src="https://user-images.githubusercontent.com/48158184/178141728-a9f9592b-d3a5-4474-8d87-620a3f859349.png">

Maybe there is more room for optimization + small refactoring?

tanhauhau commented:
> Maybe the same optimization could be applied here to increase SSR perfomance

let's keep the scope of this PR as is, can try this in a new PR if it improves the performance for SSR

MathiasWP commented:
I did a mistake and changed the "/>" to a "/" in the list of invalid attribute characters. Fixed this, so the list should match the RegEx again now.
